### ISSUE TITLE
KinematicBody3D: is_on_floor() can become false when moving against a wall

### ISSUE THREAD

**Issue by giogadi** [godotengine/godot#33833]
**Godot version:**
This issue happens in both 3.1 and in a version compiled from source (at commit 083d088de3fe7cd5d825cebebc46ce32fc87b4b0)

**OS/device including version:**
OSX Mojave 10.14.6

**Issue description:**

When a KinematicBody3D is on the ground and against a wall and running toward that wall, is_on_floor() can turn false, even though the KinematicBody3D is on the ground the whole time.

I've noticed that this issue is related to the speed of the motion (both horizontal and vertical/gravity). If both horizontal and vertical speeds are low then is_on_floor() can even be false *always* whenever running against a wall.

I did some digging into this: it seems that when vertical speed is low (for example, from one timestep of gravitational acceleration) then only one collision is reported by move_and_slide() - this collision is in the direction of the wall, not of the floor. Therefore, move_and_slide() only sees the collision with the wall direction and sets on_floor to false. However, the character still isn't falling through the floor, so the character's vertical velocity is still getting corrected; but I think both collisions are being corrected with only one motion, instead of one motion for each collision plane.

**Steps to reproduce:**
(1) Open "Godot FPS Tutorial Part 1" project.
(2) Add "print(is_on_floor())" at line 123 (right after the move_and_slide() function) to the Player's script.
(3) Run the program.
(4) Run forward directly into a wall and keep holding forward. If you check the Output, is_on_floor()'s output is flickering between true and false.

Optional step: change line 98 (the gravity update) to smaller constant velocity like "vel.y = -1.0". If you run against the wall, now is_on_floor() will always return false.

**Minimal reproduction project:**

Godot FPS Tutorial Part 1 project:
https://docs.godotengine.org/en/3.1/_downloads/d9b603ee9d6444df6b385863612e0f3f/Godot_FPS_Part_1.zip

**Comment by JamesK89** [comment]
I wanted to mention I have experienced this as well in my project. It can cause some interesting behavior on code that relies on is_on_floor since it can rapidly switch between true and false under certain circumstances.

**Godot version:**
Godot 3.1.2

**OS/device including version:**
Windows 10 Pro 1803

**Comment by RilicTheFox** [comment]
I have also experienced this. When pushing against a wall it goes from true to false every other frame, which messes with a few things in my code.

**Godot Version:**
Godot Mono 3.1.2

**OS/device including version:**
Manjaro Linux, Kernel 5.6.11

**Comment by ThePixelGuy** [comment]
Also experiencing this issue. I am assuming that it is due to the way `is_on_floor()`, `is_on_ceiling()` and `is_on_wall()` are triggered / wired in the source code (1005th line in physics_body_3d.cpp).

```
if (Math::acos(collision.normal.dot(up_direction)) <= p_floor_max_angle + FLOOR_ANGLE_THRESHOLD) { //floor

	on_floor = true;
	floor_normal = collision.normal;
	on_floor_body = collision.collider_rid;
	floor_velocity = collision.collider_vel;

	if (p_stop_on_slope) {
		if ((body_velocity_normal + up_direction).length() < 0.01 && collision.travel.length() < 1) {
			Transform gt = get_global_transform();
			gt.origin -= collision.travel.slide(up_direction);
			set_global_transform(gt);
			return Vector3();
		}
	}
} else if (Math::acos(collision.normal.dot(-up_direction)) <= p_floor_max_angle + FLOOR_ANGLE_THRESHOLD) { //ceiling
	on_ceiling = true;
} else {
	on_wall = true;
}
```

**Godot version:**
Godot 3.2.2.stable

**OS/device including version:**
Windows 10

**Comment by pouleyKetchoupp** [comment]
Can be still reproduced in 3.2.4 beta 5, but occurs only when using Bullet Physics (default settings). Switching to Godot Physics 3D fixes the issue.

**Comment by Zireael07** [comment]
@pouleyKetchoupp: Do we have a tracker for Bullet physics-only issues (as they'll likely need reporting upstream)?

**Comment by pouleyKetchoupp** [comment]
@Zireael07 Specific to Bullet doesn't necessarily mean it's fixed upstream, it's more often an issue with the Bullet physics server in Godot. Anyway, it's a good idea to have a tracker for these so I've just added one :)

**Comment by bati22** [comment]
I was able to reproduce the bug. Here are my screenshots:
Next to the wall(pushing against):
![3](https://user-images.githubusercontent.com/40314986/105403569-30f96f80-5c29-11eb-92ff-aad755093574.png)
In the bottom corner(pushing against):
![20](https://user-images.githubusercontent.com/40314986/105403786-7e75dc80-5c29-11eb-9d20-c3eadd1150cf.png)
I also looked for other cases where the function returns the wrong value.
Unfortunately I couldn't find it. In more extreme cases, when the player is in the top corner(pushing against) or between the two walls, etc. function returns valid value.
![19 1](https://user-images.githubusercontent.com/40314986/105404059-d280c100-5c29-11eb-9fbc-e58939e7d951.png)
![19 2](https://user-images.githubusercontent.com/40314986/105404068-d44a8480-5c29-11eb-8a20-e08518136578.png)
![21](https://user-images.githubusercontent.com/40314986/105404069-d6144800-5c29-11eb-8659-d71c03ed8af1.png)

**Comment by rainlizard** [comment]
This is probably the same issue so I'll post this here.
**Godot version:** v3.2.4.rc3
KinematicBody2D ``is_on_floor()`` and ``is_on_wall()`` are refusing to return true on the same frame when using small vector values.

![Untitled](https://user-images.githubusercontent.com/15337628/109800346-27741600-7c71-11eb-84d1-31e9fdfeb360.png)


```
extends KinematicBody2D
var motion = Vector2()

func _physics_process(delta):
	motion += Vector2(1,1)
	motion = move_and_slide(motion, Vector2.UP)
	
	for i in get_slide_count():
		print(get_slide_collision(i).collider.name)
	print('WALL COLLISION: ' + str(is_on_wall()))
	print('FLOOR COLLISION: ' + str(is_on_floor()))
	print('----------------------------------')
```
Prints this, flipping back and forth:
```
South Wall
WALL COLLISION: False
FLOOR COLLISION: True
----------------------------------
East Wall
WALL COLLISION: True
FLOOR COLLISION: False
----------------------------------
```

If you replace the line ``motion += Vector2(1,1)`` with ``motion += Vector2(5,5)`` they will now return true on the same frame. No idea what's going on.

Here's a minimal reproduction project:
[physics_on_wall_on_floor_bug.zip](https://github.com/godotengine/godot/files/6075655/physics_on_wall_on_floor_bug.zip)

**Comment by pouleyKetchoupp** [comment]
@rainlizard Could you please actually open a new issue for this? It's in 2D and related to a different part of the code, so it makes sense to keep it separate in this case.

**Comment by iansjolly** [comment]
> Can be still reproduced in 3.2.4 beta 5, but occurs only when using Bullet Physics (default settings). Switching to Godot Physics 3D fixes the issue.

Awesome thanks that seemed to solve my issue, no more rapid toggling between state.

**Comment by bizzk3t** [comment]
On Bullet Physics, I have a similar issue and I figured out what causes this issue in my game. However, I don't have a full solution yet. I removed my call to `linear_interpolate` and now `is_on_floor()` stays true when moving against a wall.

Right now, I am working on a solution. I will comment again if I find something. But, I thought I'd share just in case it helps anyone else.

Here is a snippet of my code for reference:

```
# is_on_floor() rapidly toggles between true and false when moving against a wall

func _physics_process(delta):

    # ...

    var hvel = velocity * Vector3(1, 0, 1)

    hvel = hvel.linear_interpolate(direction * speed, acceleration * delta)

    velocity.x = hvel.x
    velocity.z = hvel.z

    velocity.y -= gravity * delta

    velocity = move_and_slide(velocity, Vector3.UP)



# is_on_floor() is true when moving against wall
# However, movements are quick and unnatural

func _physics_process(delta):

    # ...

    var hvel = Vector3(1, 0, 1) * direction * speed

    velocity.x = hvel.x
    velocity.z = hvel.z

    velocity.y -= gravity * delta

    velocity = move_and_slide(velocity, Vector3.UP)

```

**Comment by Calinou** [comment]
See [Shfty's character movement manifesto](https://markdownpastebin.com/?id=d9d61e67f9d64db2bd215f165b931449) for a workaround:

> - Unstable floor results make my 'is grounded' flag jitter
>   - This seems to be down to the collision margin system
>   - Use a rolling window (2-3 frames should be sufficient) that folds down to a single bool using OR to de-noise the floor check

**Comment by akien-mga** [comment]
Still reproducible in 3.5 RC 1 and 5d9c7eaef9a0d572fafa467356a3f611ada50986.

**Comment by return2monki** [comment]
Hi I noticed this same problem when debugging in 3.5 yesterday. is_on_floor() returns false most of the time when moving into a corner. I tried to get my "moving" variable to return as false when moving into a wall even when using velocity.length() > 4 I can't get it to work perfectly. Visually the character does not move at all but the Y variable fluctuates **heavily**.

Some workarounds that I tried:

Setting axis_lock_y(2, true) to lock your characters y position and then setting it to false when you need to jump or move upwards, did NOT work as it still keeps saying the Y velocity is changing and is_on_floor returns false or at least rapidly changes.

The following one actually fixed my specific problem with the moving variable when running into corner, but it does not fix the is_on_floor:

```
	var pos = translation
	...

	if str(pos.length_squared()).pad_decimals(3) != str(translation.length_squared()).pad_decimals(3):
		moving = true
		pos = translation
	elif str(pos.length_squared()).pad_decimals(3) == str(translation.length_squared()).pad_decimals(3):
		moving = false
```

I can only get the result that I want by doing this specific comparison, this actually fixes my problem instead of doing the preferred option which is just checking velocity.length() as the y velocity acts really strange in corners. Using global_translation was giving me an error about instancing so I'm using translation.

**So in short: comparing your actual position directly works for what I'm doing as you don't actually change position, but for some reason the Y velocity does change which is the entire problem with is_on_floor() I believe.**

Running into a wall doesn't give problems, it has to be a corner specifically.

**Comment by Calinou** [comment]
> I just checked and Godot physics fixes all these problems, will it be default in Godot 4?

Yes :slightly_smiling_face:
